sopel/db.py

```diff
@@ -39,51 +84,65 @@ class SopelDB(object):
     directory as the config."""
 
     def __init__(self, config):
-        path = config.core.db_filename
-        config_dir, config_file = os.path.split(config.filename)
-        config_name, _ = os.path.splitext(config_file)
-        if path is None:
-            path = os.path.join(config_dir, config_name + '.db')
-        path = os.path.expanduser(path)
-        if not os.path.isabs(path):
-            path = os.path.normpath(os.path.join(config_dir, path))
-        self.filename = path
-        self._create()
+        # MySQL - mysql://username:password@localhost/db
+        # SQLite - sqlite:////home/sopel/.sopel/default.db
+        db_type = config.core.db_type
+
+        if db_type and db_type == 'mysql':
```

Comment:
This ties the database to mysql and sqlite only. Can we first check for sqlite and otherwise build the connection string from the other args?

Also, shouldn't db_type be required? This should never be None in my opinion. Does the configuration support required fields?
```python
# some imports first, obviously not intended for literally right here...
from sqlalchemy.engine.url import URL

if not db_type:
    raise Exception("Please make sure core.db_type is defined")
if db_type == 'sqlite':
    # sqlite section here...
    pass
else:
    db_user = config.core.db_user
    db_pass = config.core.db_pass
    db_host = config.core.db_host
    db_port = config.core.db_port  # this is missing, this should totes exist, and a default None value is fine
    db_database = config.core.db_database
    db_query = config.core.db_query  # this is an optional set of options to be passed, probably not important enough)

    self.engine = create_engine(URL(db_type, db_user, db_pass, db_host, db_port, db_database, db_query))
```

Reply:
Configuration settings that have no default set will throw an error when code tries to use them:

https://github.com/sopel-irc/sopel/blob/8a6df765072e44ecbfb9acdcc3275fcd2cd4284f/sopel/config/types.py#L145-L151

Reply:
> `db_type` should have `default='sqlite'` in the config class. There's no reason to do a hard require if a sensible default will do (and SQLite will remain the default engine, so this is fine).

And a defaulted filename? This would just get dropped/ignored with using anything other than sqlite.

Reply:
I haven't looked through this code yet (in the middle of a busy few days), but the current state is a default filename based on the config name if not set:

https://github.com/sopel-irc/sopel/blob/a0c567c09083414d20ad4f989012f80062d6d2fa/sopel/db.py#L42-L46